src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs

```diff
@@ -35,11 +35,21 @@ public SortedDictionary(IDictionary<TKey, TValue> dictionary, IComparer<TKey>? c
                 throw new ArgumentNullException(nameof(dictionary));
             }
 
-            _set = new TreeSet<KeyValuePair<TKey, TValue>>(new KeyValuePairComparer(comparer));
+            var keyValuePairComparer = new KeyValuePairComparer(comparer);
 
-            foreach (KeyValuePair<TKey, TValue> pair in dictionary)
+            if (dictionary is SortedDictionary<TKey, TValue> sortedDictionary &&
+                sortedDictionary._set.Comparer.Equals(keyValuePairComparer))
```

Comment:
Will this equality check ever evaluate to `true`? You're comparing with a brand new `KeyValuePairComparer` instance which AFAIK has reference equality semantics.

Reply:
For instace, here's a bit of code I tested in C# interactive:
```csharp
> var comparer = Comparer<int>.Default;
> var kvpc1 = new SortedDictionary<int, int>.KeyValuePairComparer(comparer);
> var kvpc2 = new SortedDictionary<int, int>.KeyValuePairComparer(comparer);
> kvpc1.Equals(kvpc2)
false
```

Reply:
Oh nevermind, I just saw you actually changed equality for the wrapper class.